scalafix-core/shared/src/main/scala/scalafix/internal/util/LazySymbolTable.scala

```diff
@@ -0,0 +1,117 @@
+import scala.collection.concurrent.TrieMap
+import scala.meta.internal.semanticdb3.Scala._
+import scala.meta.internal.{semanticdb3 => s}
+import scala.util.control.NonFatal
+
+/**
+  * Implementation of SymbolTable that lazily loads symbols on demand.
+  *
+  * @param mclasspath The classpath to load symbols from. Needs to be pre-processed by metacp,
+  *                   see ClasspathOps.toMclasspath in scalafix-cli.
+  */
+class LazySymbolTable(mclasspath: Classpath) extends SymbolTable {
+
+  // Map from symbols to the paths to their corresponding semanticdb file where they are stored.
+  private val unloadedSymbols = TrieMap.empty[String, AbsolutePath]
+  private val loadedSymbols = TrieMap.empty[String, s.SymbolInformation]
+  private val semanticdb = "META-INF/semanticdb"
+  private val semanticIdx = "META-INF/semanticdb.semanticidx"
+
+  mclasspath.shallow.foreach(loadSemanticdbIndex)
+
+  override def info(symbol: String): Option[s.SymbolInformation] = {
+    var result = loadedSymbols.get(symbol)
+    if (result.isEmpty) {
+      loadSymbolFromClasspath(symbol)
+      result = loadedSymbols.get(symbol)
+      if (result.isEmpty && symbol.owner != Symbols.None) {
+        info(symbol.owner)
+        result = loadedSymbols.get(symbol)
+      }
+    }
+    result
+  }
+
+  private def loadSymbolFromClasspath(symbol: String): Unit = {
+    val toLoad = unloadedSymbols.get(symbol)
+    if (toLoad.isDefined) {
+      val in = toLoad.get.toNIO.toUri.toURL.openStream()
+      val semanticdb =
+        try s.TextDocuments.parseFrom(in)
+        finally in.close()
+      semanticdb.documents.foreach { document =>
+        document.symbols.foreach { info =>
+          loadedSymbols.put(info.symbol, info)
+        }
+      }
+      unloadedSymbols.remove(symbol)
+    }
+  }
+
+  private def loadIndex(root: AbsolutePath): Unit = {
+    val path = root.resolve("META-INF").resolve("semanticdb.semanticidx")
+    if (!Files.isRegularFile(path.toNIO)) return
+    val in = Files.newInputStream(path.toNIO)
+    val index =
+      try s.Index.parseFrom(in)
+      finally in.close()
+    index.toplevels.foreach { toplevel =>
+      unloadedSymbols.put(
+        toplevel.symbol,
+        root.resolve(semanticdb).resolve(toplevel.uri))
+    }
+  }
+
+  private def loadSemanticdbIndex(root: AbsolutePath): Unit = {
+    if (root.isDirectory) {
+      loadIndex(root)
+    } else if (PathIO.extension(root.toNIO) == "jar") {
+      withJarFileSystem(root)(loadIndex)
+    } else {
+      throw new IllegalArgumentException(root.toString())
+    }
+  }
+
+  private def withJarFileSystem[T](path: AbsolutePath)(
+      f: AbsolutePath => T): T = {
+    val fs = newJarFileSystem(path)
+    // NOTE(olafur): We don't fs.close() because that can affect another place where `FileSystems.getFileSystems`
+    // was used due to a `FileSystemAlreadyExistsException`. I don't know what the best solution is for reading the
+    // same zip file from multiple concurrent places.
```

Comment:
Maybe this is/was the source of deadlocks?

Reply:
That was my first suspicion as well! I continuously ran the test suite to see problems. This wasn't the cause of the deadlock, I checked the stack traces and  they led to metacp.